File: app/src/main/java/me/echeung/moemoekyun/client/network/NetworkClient.java
package me.echeung.moemoekyun.client.network;

import me.echeung.moemoekyun.util.system.NetworkUtil;
import okhttp3.OkHttpClient;
import okhttp3.Request;

public class NetworkClient {

    private static final String HEADER_CONTENT_TYPE = "Content-Type";
    private static final String CONTENT_TYPE = "application/json";

    private static final String HEADER_ACCEPT = "Accept";
    private static final String ACCEPT = "application/vnd.listen.v4+json";

    private static final String HEADER_USER_AGENT = "User-Agent";

    public static OkHttpClient getClient() {
        return new OkHttpClient.Builder()
                .addNetworkInterceptor(chain -> {
                    final Request request = chain.request();

                    final Request newRequest = request.newBuilder()
                            .addHeader(HEADER_USER_AGENT, NetworkUtil.getUserAgent())
                            .addHeader(HEADER_CONTENT_TYPE, CONTENT_TYPE)
                            .addHeader(HEADER_ACCEPT, ACCEPT)
                            .build();

                    return chain.proceed(newRequest);
                })
                .build();
    }

}

File: app/src/main/kotlin/me/echeung/moemoekyun/client/network/NetworkClient.kt
package me.echeung.moemoekyun.client.network

import android.content.Context
import me.echeung.moemoekyun.client.auth.AuthUtil
import me.echeung.moemoekyun.util.system.NetworkUtil
import okhttp3.OkHttpClient
import java.io.File
import java.util.concurrent.TimeUnit

class NetworkClient(
    context: Context,
    authUtil: AuthUtil,
) {

    val client = OkHttpClient.Builder()
        .addNetworkInterceptor { chain ->
            val request = chain.request()

            val newRequest = request.newBuilder()
                .header("User-Agent", NetworkUtil.userAgent)
                .header("Content-Type", "application/json")

            // MFA login
            if (authUtil.mfaToken != null) {
                newRequest.header(HEADER_AUTHZ, authUtil.mfaAuthTokenWithPrefix)
            }

            // Authorized calls
            if (authUtil.isAuthenticated) {
                newRequest.header(HEADER_AUTHZ, authUtil.authTokenWithPrefix)
            }

            chain.proceed(newRequest.build())
        }
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    val apolloCache = File(context.externalCacheDir, "apolloCache")
}

private const val HEADER_AUTHZ = "Authorization"
